concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs

```diff
@@ -490,37 +562,65 @@ instance (MonadLogger (PersistentTreeStateMonad bs m),
+
+    getRecentBlockStatus bh = do
+      st <- use (blockTable . liveMap . at' bh)
+      case st of
+        Just (BlockAlive bp) -> return $ TS.RecentBlock (TS.BlockAlive bp)
+        Just (BlockPending bp) -> return $ TS.RecentBlock (TS.BlockPending bp)
+        Nothing -> do
+            lf <- use lastFinalized
+            if bh == bpHash lf then do
+              lfr <- use lastFinalizationRecord
+              return $ TS.RecentBlock (TS.BlockFinalized lf lfr)
+            else do
+              b <- memberBlockTable bh
+              if b then
+                return TS.OlderThanLastFinalized
+              else do
+                  deadBlocks <- use (blockTable . deadCache)
+                  return $! if memberDeadCache bh deadBlocks then TS.RecentBlock TS.BlockDead else TS.Unknown
+
     makeLiveBlock block parent lastFin st arrTime energy = do
             blockP <- makePersistentBlockPointerFromPendingBlock block parent lastFin st arrTime energy
-            blockTable . at' (getHash block) ?=! BlockAlive blockP
+            blockTable . liveMap . at' (getHash block) ?=! BlockAlive blockP
             return blockP
-    markDead bh = blockTable . at' bh ?= BlockDead
+    markDead bh = do
+            blockTable . liveMap . at' bh .=! Nothing
+            blockTable . deadCache %=! insertDeadCache bh
     type MarkFin (PersistentTreeStateMonad bs m) = Maybe (StoredBlock (MPV m) (BlockStatePointer bs))
-    markFinalized bh fr = use (blockTable . at' bh) >>= \case
+    markFinalized bh fr = use (blockTable . liveMap . at' bh) >>= \case
             Just (BlockAlive bp) -> do
               st <- saveBlockState (_bpState bp)
-              blockTable . at' bh ?=! BlockFinalized (finalizationIndex fr)
+              -- NB: This only makes sense if no block lookups are done between
+              -- the call to this function and 'wrapUpFinalization'. This is
+              -- currently the case, and must remain the case in the future.
+              -- That is, finalization must remain atomic, or this handling, and
+              -- wrapUpFinalization must be changed.
+              blockTable . liveMap . at' bh .=! Nothing
               return $ Just StoredBlock{
                   sbFinalizationIndex = finalizationIndex fr,
                   sbInfo = _bpInfo bp,
                   sbBlock = _bpBlock bp,
                   sbState = st
                 }
             _ -> return Nothing
-    markPending pb = blockTable . at' (getHash pb) ?=! BlockPending pb
-    markAllNonFinalizedDead = blockTable %=! fmap nonFinDead
-        where
-            nonFinDead BlockAlive{} = BlockDead
-            nonFinDead BlockPending{} = BlockDead
-            nonFinDead o = o
+    markPending pb = blockTable . liveMap . at' (getHash pb) ?=! BlockPending pb
+    markAllNonFinalizedDead = do
+      blockTable . liveMap .=! HM.empty
+      blockTable . deadCache .=! emptyDeadCache
```

Comment:
This behaviour seems not at all in the spirit of how the dead cache should operate. This is probably fine for the context the function is used in, but the documentation (and possibly the function name) should be updated to reflect this.

Reply:
I updated the comment to reduce the scope of the function (and renamed it).